Sources/Document/CRDT/CRDTRoot.swift

```diff
@@ -0,0 +1,251 @@
+
+/**
+ * `CRDTRoot` is a structure represents the root. It has a hash table of
+ * all elements to find a specific element when applying remote changes
+ * received from server.
+ *
+ * Every element has a unique time ticket at creation, which allows us to find
+ * a particular element.
+ */
+class CRDTRoot {
+    private var rootObject: CRDTObject
+    private var elementPairMapByCreatedAt: [TimeTicket: CRDTElementPair] = [:]
+    private var removedElementSetByCreatedAt: Set<TimeTicket> = Set()
+    private var textWithGarbageSetByCreatedAt: Set<TimeTicket> = Set()
+
+    init(rootObject: CRDTObject = CRDTObject(createdAt: TimeTicket.initial)) {
+        self.rootObject = rootObject
+        self.elementPairMapByCreatedAt[self.rootObject.getCreatedAt()] = (element: self.rootObject, parent: nil)
+
+        self.rootObject.getDescendants(callback: { element, parent in
+            self.registerElement(element: element, parent: parent)
+            return false
+        })
+    }
+
+    /**
+     * `find` returns the element of given creation time.
+     */
+    func find(createdAt: TimeTicket) -> CRDTElement? {
+        return self.elementPairMapByCreatedAt[createdAt]?.element
+    }
+
+    private let subPathPrefix = "$"
+    private let subPathSeparator = "."
+
+    /**
+     * `createSubPaths` creates an array of the sub paths for the given element.
+     */
+    func createSubPaths(createdAt: TimeTicket) throws -> [String] {
+        guard let pair = self.elementPairMapByCreatedAt[createdAt] else {
+            return []
+        }
+
+        var result: [String] = []
+        var pairForLoop: CRDTElementPair = pair
+        while let parent = pairForLoop.parent {
+            let createdAt = pairForLoop.element.getCreatedAt()
+            var subPath = try parent.subPath(createdAt: createdAt)
+            subPath = self.escapeSubpath(subPath)
+            result.append(subPath)
+            guard let parentPair = self.elementPairMapByCreatedAt[parent.getCreatedAt()] else {
+                break
+            }
+
+            pairForLoop = parentPair
+        }
+
+        result.append(self.subPathPrefix)
+        return result.reversed()
+    }
+
+    private func escapeSubpath(_ target: String) -> String {
+        return [self.subPathPrefix, self.subPathSeparator].reduce(target) { partialResult, seq in
+            partialResult.replacingOccurrences(of: seq, with: "\\\(seq)")
+        }
+    }
+
+    /**
+     * `createPath` creates path of the given element.
+     */
+    func createPath(createdAt: TimeTicket) throws -> String {
+        return try self.createSubPaths(createdAt: createdAt).joined(separator: self.subPathSeparator)
+    }
+
+    /**
+     * `registerElement` registers the given element to hash table.
+     */
+    func registerElement(element: CRDTElement, parent: CRDTContainer?) {
+        self.elementPairMapByCreatedAt[element.getCreatedAt()] = (element, parent)
```

Comment:
@hackerwins In js-sdk, the order of parent and element is as follows. 

```
  public registerElement(element: CRDTElement, parent: CRDTContainer): void {
    this.elementPairMapByCreatedAt.set(element.getCreatedAt().toIDString(), {
      parent,
      element,
    });
  }
```
Is that right?

Reply:
Thanks for your report. There seems to be a problem.
I'll check a little more in JS SDK.

Reply:
Thanks for your report. 👍 

The following are shorthand properties in JavaScript(ES6).

```js
    {
      parent,
      element,
    }
```

Same as the below code:

```js
    {
      parent: parent,
      element: element,
    }
```

https://ui.dev/shorthand-properties